libs/arcade-tdk/arcade_tdk/providers/http/error_adapter.py

```diff
@@ -0,0 +1,199 @@
+import logging
+from datetime import datetime, timezone
+from typing import Any
+from urllib.parse import urlparse
+
+from arcade_core.errors import (
+    UpstreamError,
+    UpstreamRateLimitError,
+)
+
+logger = logging.getLogger(__name__)
+
+RATE_HEADERS = ("retry-after", "x-ratelimit-reset", "x-ratelimit-reset-ms")
+
+
+class BaseHTTPErrorMapper:
+    """Base class for HTTP error mapping functionality."""
+
+    def _parse_retry_ms(self, headers: dict[str, str]) -> int:
+        """
+        Parses a rate limit header and returns the number
+        of milliseconds until the rate limit resets.
+
+        Args:
+            headers: A dictionary of HTTP headers.
+
+        Returns:
+            The number of milliseconds until the rate limit resets.
+            Defaults to 1000ms if a rate limit header is not found or cannot be parsed.
+        """
+        val = next((headers.get(h) for h in RATE_HEADERS if headers.get(h)), None)
+        # No rate limit header found
+        if val is None:
+            return 1_000
+        # Rate limit header is a number of seconds
+        if val.isdigit():
+            key = next((h for h in RATE_HEADERS if headers.get(h) == val), "")
+            if key.endswith("ms"):
+                return int(val)
+            return int(val) * 1_000
+        # Rate limit header is an absolute date
+        try:
+            dt = datetime.strptime(val, "%a, %d %b %Y %H:%M:%S %Z")
+            return int((dt - datetime.now(timezone.utc)).total_seconds() * 1_000)
+        except Exception:
+            return 1_000
```

Comment:
Perhaps would be good to log the header parsing errors. This way we can find out non-standard rate-limit headers and extend our implementation.